File: read-only-state.ts
type Reader<E, A> = (e: E) => A;

function runReader<E, A>(f: Reader<E, A>, e: E): A {
  return f(e);
}

const runReaderDup: <E, A>(f: Reader<E, A>, e: E)=> A  = (f, e) => f(e);

// pure / return — создает Reader, игнорирующий окружение
export const of = <E, A>(value: A): Reader<E, A> => (_env: E) => value;

// fmap
export const map = <E, A, B>(f: (a: A) => B) =>
  (ra: Reader<E, A>): Reader<E, B> =>
    (env: E) => f(ra(env));

// ap — для Applicative
export const ap = <E, A, B>(rf: Reader<E, (a: A) => B>) =>
  (ra: Reader<E, A>): Reader<E, B> =>
    (env: E) => rf(env)(ra(env));

// flatMap / bind / >>=
export const flatMap = <E, A, B>(f: (a: A) => Reader<E, B>) =>
  (ra: Reader<E, A>): Reader<E, B> =>
    (env: E) => f(ra(env))(env);

// ask — возвращает окружение как есть
export const ask = <E>(): Reader<E, E> =>
  (env: E) => env;

// -------------------- Пример --------------------

const example: Reader<string, string> =
  flatMap<string, string, string>((env) => of(`Hello, ${env}!`))(ask());

// Проверка
console.log(runReader(example, "Leonty")); // → Hello, Leonty!

File: read-only-state.hs
module Main where

newtype Reader e a = Reader (e -> a)

runReader :: Reader e a -> e -> a
runReader (Reader f) e = f e

instance Functor (Reader e) where
  fmap f (Reader g) = Reader (f . g)

instance Applicative (Reader e) where
  pure x = Reader (\_ -> x)
  Reader f <*> Reader g = Reader (\e -> f e (g e))

instance Monad (Reader e) where
  ra >>= k = Reader (\e -> runReader (k (runReader ra e)) e)
  -- return x = Reader (\e -> x)
  
ask :: Reader e e
ask = Reader id
  
example :: Reader String String
example = do
  env <- ask
  return ("Hello, " ++ env ++ "!")

main :: IO ()
main = putStrLn (runReader example "Leonty")